### PR TITLE
IGVF-1039-audits-page

### PR DESCRIPTION
**Pull request by khzlin**

### CODE REVIEW

**Review by forresttanaka — CHANGES_REQUESTED** [review]
For the “Audit Level” column, I feel like the audit icons should come before the level text so that all the icons line up on the left and all the text lines up to their right — “{icon} Warning” instead of “Warning {icon}.” Right now the icons zigzag left and right down the page.

File: components/audit-dynamic-table.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            );
+          })}
+        </div>
+      );
+    },
+  },
+  {
+    id: "audit_category",
+    title: "Audit Category",
+    display: ({ source }) => `${source.audit_category}`,
```
[the remark this answers is not in the document]
Comment (khzlin) [reply components/audit-dynamic-table.js:37]: done

File: components/audit-dynamic-table.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                <div className="flex items-center justify-start gap-1">
+                  <div>{mapping.humanReadable}</div>
+                  <mapping.Icon className={`h-4 w-4 ${mapping.color}`} />
+                </div>
+              </Fragment>
+            );
+          })}
+        </div>
+      );
+    },
```
[the remark this answers is not in the document]
Comment (khzlin) [reply components/audit-dynamic-table.js:32]: not sorting

File: components/audit-dynamic-table.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    <DataGridContainer>
+      <SortableGrid data={data} columns={auditColumns} />
+    </DataGridContainer>
+  );
+}
+
+AuditDynamicTable.propTypes = {
+  // Audits to display
+  data: PropTypes.arrayOf(PropTypes.object).isRequired,
+};
```
[the remark this answers is not in the document]
Comment (khzlin) [reply components/audit-dynamic-table.js:61]: done

File: components/navigation.js [hunk 4]
```diff
@@ -660,6 +660,14 @@ function NavigationExpanded({ navigationClick, toggleNavCollapsed }) {
...
             Schemas
           </NavigationHrefItem>
+          <NavigationHrefItem
+            id="audits"
+            href="/audits"
+            navigationClick={navigationClick}
+            isChildItem
+          >
+            Audits
+          </NavigationHrefItem>
```
[the remark this answers is not in the document]
Comment (khzlin) [reply components/navigation.js:670]: we're going with Audit Documentation

File: lib/__tests__/general.test.js [hunk 5]
```diff
@@ -130,3 +131,11 @@ describe("Test the truncateJson utility function", () => {
     expect(truncated).toStrictEqual(`{"a":[1,2,3,4,5,6,7,...`);
   });
 });
+
+describe("Test the formatPropertyName utility function", () => {
+  it("Should convert a string to formatPropertyName case", () => {
+    expect(formatPropertyName("Property_Name")).toBe("Property Name");
+    expect(formatPropertyName("PRoperty_NamE")).toBe("Property Name");
+    expect(formatPropertyName("property_name")).toBe("Property Name");
```
[the remark this answers is not in the document]
Comment (khzlin) [reply lib/__tests__/general.test.js:139]: done

File: lib/general.ts [hunk 6]
```diff
@@ -148,3 +148,16 @@ export function truncateJson(
     ? `${json.substring(0, maxOutputLength)}...`
     : json;
 }
+
+/**
+ * Convert arbitrary text to a human readable string. All underscores gets replaced with a space and the first letter of each word gets capitalized.
+ * @param {string} text Text to convert.
+ * @returns {string} `text` converted to shishkebab case.
```
[the remark this answers is not in the document]
Comment (khzlin) [reply lib/general.ts:155]: done

File: lib/general.ts [hunk 7]
```diff
@@ -148,3 +148,16 @@ export function truncateJson(
...
+ */
+export function formatPropertyName(text: string): string {
```
[the remark this answers is not in the document]
Comment (khzlin) [reply lib/general.ts:157]: done

File: pages/audits.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import Markdown from "../components/markdown";
+// lib
+import errorObjectToProps from "../lib/errors";
+import FetchRequest from "../lib/fetch-request";
+import { formatPropertyName } from "../lib/general";
+
+export default function AuditDoc({ auditArrayObject }) {
+  return (
+    <>
+      <Breadcrumbs />
```
[the remark this answers is not in the document]
Comment (khzlin) [reply pages/audits.js:16]: done

File: lib/general.ts [hunk 9]
```diff
@@ -148,3 +148,16 @@ export function truncateJson(
     ? `${json.substring(0, maxOutputLength)}...`
     : json;
 }
+
+/**
+ * Convert arbitrary text to a human readable string. All underscores gets replaced with a space and the first letter of each word gets capitalized.
```
[the remark this answers is not in the document]
Comment (khzlin) [reply lib/general.ts:153]: done

File: pages/audits.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      <h1 className="tight my-4 text-center text-xl font-bold text-black dark:text-gray-300 sm:text-2xl md:my-10 md:text-3xl lg:text-4xl">
+        Audits
+      </h1>
+      <Markdown
+        markdown={
+          "The IGVF Data and Administration Coordinating Center has established an audit system, utilizing flags, to detect discrepancies in the data. While dependencies ensure metadata accuracy within an individual object, audits primarily focus on validating metadata between linked objects. For example, a specific audit has been implemented to verify the linkage of Biosamples with donors sharing the same taxa. Within each audit category, there could be one or more icons, each assigned a distinct color corresponding to the severity level of the audit category."
+        }
+      />
```
[the remark this answers is not in the document]
Comment (khzlin) [reply pages/audits.js:24]: done

File: pages/audits.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+      {Object.keys(auditArrayObject).map((itemType) => {
+        const typeAudits = auditArrayObject[itemType];
+        const filteredAudits = typeAudits.filter(
+          (audit) => audit.audit_levels[0] !== "INTERNAL_ACTION"
+        );
+        if (filteredAudits.length > 0) {
+          return (
+            <>
+              <h2 className="mb-1 px-2 pt-8 text-lg font-semibold text-brand">
```
[the remark this answers is not in the document]
Comment (khzlin) [reply pages/audits.js:33]: done

File: pages/audits.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+          );
+        }
+        return null;
+      })}
+    </>
+  );
+}
+
+AuditDoc.propTypes = {
+  auditArrayObject: PropTypes.object.isRequired,
```
[the remark this answers is not in the document]
Comment (khzlin) [reply pages/audits.js:47]: done

File: pages/audits.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        ...auditDoc[key],
+      };
+    });
+    const auditArrayObject = _.groupBy(auditArrayVersion, "key");
+    return {
+      props: {
+        auditDoc,
+        auditArrayVersion,
+        auditArrayObject,
+      },
```
[the remark this answers is not in the document]
Comment (khzlin) [reply pages/audits.js:66]: done

File: pages/audits.js [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        color corresponding to the severity level of the audit category.
+      </p>
+      {Object.keys(auditsGroupedByCollection).map((itemType) => {
+        const typeAudits = auditsGroupedByCollection[itemType];
+        const filteredAudits = typeAudits.filter(
+          (audit) => audit.audit_levels[0] !== "INTERNAL_ACTION"
+        );
+        if (filteredAudits.length > 0) {
+          return (
+            <>
```
Comment (forresttanaka) [inline pages/audits.js:40]: During run time, getting a missing key error. Need to change the `<>` to `<Fragment key={itemType}>`.

**Review by forresttanaka — CHANGES_REQUESTED** [review]
Looking nice! Just have to fix a run-time key error that shows up in the debugger console. Easy fix.

File: pages/audits.js (hunk 14 above)
Comment (khzlin) [reply pages/audits.js:40]: @forresttanaka` itemType` isn't defined until later in L33 so can i do `<Fragment key={keyedAudits}>`?

**Comment by forresttanaka** [comment]
@khzlin Also, take a look at the audits page in dark mode. I think the h2 titles are a _tad_ dark against the dark background, though not too bad. If you agree, you just need to add a class "dark:text-something". You can choose a color from this table, so for example, if you choose a slate color with a 300 shade (not a suggestion), the class would be `dark:text-slate-300`. You can also hardcode a hex color with something like `dark:text-[#328d12]`.

File: pages/audits.js (hunk 14 above)
Comment (forresttanaka) [reply pages/audits.js:40]: @khzlin You need to change lines 40 and 45, so `itemType` exists at that point.

File: pages/audits.js [hunk 15]
```diff
@@ -37,12 +38,12 @@ export default function AuditDoc({ auditDoc }) {
         );
         if (filteredAudits.length > 0) {
           return (
-            <>
-              <h2 className="mb-1 mt-8 text-lg font-semibold text-brand">
+            <Fragment key={keyedAudits}>
```
Comment (forresttanaka) [inline pages/audits.js:41]: The key has to be a number or string, but `keyedAudits` is an array, so I see lots of big errors in the browser console now. Use `itemType` which is a string that’s unique for every iteration.

**Review by forresttanaka — CHANGES_REQUESTED** [review]

File: pages/audits.js (hunk 15 above)
Comment (khzlin) [reply pages/audits.js:41]: oh sorry. I thought i had changed it following our conversation. Sorry about that!

**Review by forresttanaka — APPROVED** [review]

**Pull request merged**